
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>configuration file format &#8212; PISCES 0.1.1 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="configuration-file-format">
<h1>configuration file format<a class="headerlink" href="#configuration-file-format" title="Permalink to this headline">Â¶</a></h1>
<p>PISCES is driven by JSON configuration files. The key:value
pairs in this configuration file then are used by all subsequent <code class="docutils literal notranslate"><span class="pre">pisces</span></code>
subcommands. Paths can be
either local files, or HTTP/FTP URLs. Notice that chimeric organisms such as a human/mouse xenograft organisms are
easily defined by specifying each transcriptome as a separate GTF/FASTA pair.
Example transcriptomes for human, mouse, and human/mouse xenograft are included
in the default PISCES distribution. If you need to support another organism, this
can easily be accomplished by adding your transcriptome of interest in a new
configuration file. The configuration file format follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;human&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;downloads&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces/downloads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index_dir&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces&quot;</span><span class="p">,</span>
        <span class="s2">&quot;default&quot;</span><span class="p">:</span> <span class="s2">&quot;gencode_basic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;gencode_basic&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;gtfs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.basic.annotation.gtf.gz&quot;</span><span class="p">],</span>
                <span class="s2">&quot;fastas&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz&quot;</span><span class="p">],</span>
                <span class="s2">&quot;extra_fastas&quot;</span><span class="p">:</span> <span class="p">[],</span>
                <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;-k&quot;</span><span class="p">:</span> <span class="mi">31</span><span class="p">,</span>
                    <span class="s2">&quot;infer_features&quot;</span><span class="p">:</span> <span class="n">false</span>
                <span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">},</span>
    <span class="s2">&quot;mouse&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;downloads&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces/downloads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index_dir&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces&quot;</span><span class="p">,</span>
        <span class="s2">&quot;default&quot;</span><span class="p">:</span> <span class="s2">&quot;gencode_basic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;gencode_basic&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;gtfs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.basic.annotation.gtf.gz&quot;</span><span class="p">],</span>
                <span class="s2">&quot;fastas&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz&quot;</span><span class="p">],</span>
                <span class="s2">&quot;extra_fastas&quot;</span><span class="p">:</span> <span class="p">[],</span>
                <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;-k&quot;</span><span class="p">:</span> <span class="mi">31</span><span class="p">,</span>
                    <span class="s2">&quot;infer_features&quot;</span><span class="p">:</span> <span class="n">false</span>
                <span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">},</span>
    <span class="s2">&quot;human-mouse&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;downloads&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces/downloads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index_dir&quot;</span><span class="p">:</span> <span class="s2">&quot;$HOME/pisces&quot;</span><span class="p">,</span>
        <span class="s2">&quot;default&quot;</span><span class="p">:</span> <span class="s2">&quot;gencode_basic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;gencode_basic&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;gtfs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.basic.annotation.gtf.gz&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.basic.annotation.gtf.gz&quot;</span>
                <span class="p">],</span>
                <span class="s2">&quot;fastas&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz&quot;</span><span class="p">],</span>
                <span class="s2">&quot;extra_fastas&quot;</span><span class="p">:</span> <span class="p">[],</span>
                <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;-k&quot;</span><span class="p">:</span> <span class="mi">31</span><span class="p">,</span>
                    <span class="s2">&quot;infer_features&quot;</span><span class="p">:</span> <span class="n">false</span>
                <span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
</div>


          </div>
          
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Novartis Institutes for Biomedical Research.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/config.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>